File: backend/app/schemas/tenderly.py
from typing import Literal

from pydantic import BaseModel, Field


class TenderlySimulationRequest(BaseModel):
    action_label: str = Field(min_length=4, max_length=120)
    protocol: str = Field(min_length=2, max_length=64)
    token_symbol: str = Field(min_length=2, max_length=12)
    amount_usd: float = Field(gt=0, le=5_000_000)
    approval_scope: Literal["none", "limited", "unlimited"] = "none"
    approval_duration_hours: int = Field(ge=0, le=168)
    target: str = Field(min_length=2, max_length=64)


class TokenMovementPreview(BaseModel):
    token: str
    direction: Literal["in", "out"]
    amount_usd: float = Field(ge=0)


class TenderlySimulationResponse(BaseModel):
    deterministic: bool = True
    token_movements: list[TokenMovementPreview]
    gas_estimate_usd: float = Field(ge=0)
    approval_size_usd: float = Field(ge=0)
    transaction_failure_risk: Literal["low", "medium", "high"]
    suspicious_behavior_detected: bool
    human_explanation: str = Field(min_length=20, max_length=320)

File: frontend/lib/api/tenderly.ts
import { apiClient } from "@/lib/api/client";

export type TenderlySimulationRequest = {
  action_label: string;
  protocol: string;
  token_symbol: string;
  amount_usd: number;
  approval_scope: "none" | "limited" | "unlimited";
  approval_duration_hours: number;
  target: string;
};

export type TokenMovementPreview = {
  token: string;
  direction: "in" | "out";
  amount_usd: number;
};

export type TenderlySimulationResponse = {
  deterministic: boolean;
  token_movements: TokenMovementPreview[];
  gas_estimate_usd: number;
  approval_size_usd: number;
  transaction_failure_risk: "low" | "medium" | "high";
  suspicious_behavior_detected: boolean;
  human_explanation: string;
};

export async function simulateTransaction(
  payload: TenderlySimulationRequest
): Promise<TenderlySimulationResponse> {
  return apiClient<TenderlySimulationResponse>({
    path: "/api/v1/agent/tenderly/simulate",
    method: "POST",
    body: JSON.stringify(payload)
  });
}
